pkg/sos/client.go

```diff
@@ -0,0 +1,140 @@
+// Copyright 2018 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package sos
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net"
+	"net/http"
+	"os"
+	"os/signal"
+	"strconv"
+	"strings"
+	"syscall"
+
+	"github.com/gorilla/mux"
+)
+
+// RegistersNecessaryPatterns registers all the neccesary patterns needed
+// to make a service becomes a SoS client.
+func RegistersNecessaryPatterns(router *mux.Router) {
+	router.HandleFunc("/ping", func(w http.ResponseWriter, r *http.Request) {
+		w.Write([]byte("pong"))
+	}).Methods("GET")
+}
+
+// RegisterServiceWithSos tries to register a service with SoS.
+// If an non-nil error is returned, the service needs to exit immediately.
+func RegisterServiceWithSos(service string, port uint) error {
+	return registerServiceWithSos(service, port, "http://localhost:"+PortNum)
+}
+
+func registerServiceWithSos(service string, port uint, sosServerURL string) error {
+	m := RegisterReqJson{service, port}
+	return makeRequestToServer("POST", sosServerURL+"/register", m)
+}
+
+// UnregisterServiceWithSos makes a request to SoS Server to unregister the service.
+// This function should be called before a service exit.
+func UnregisterServiceWithSos(service string) error {
+	return unregisterServiceWithSos(service, "http://localhost:"+PortNum)
+}
+
+func unregisterServiceWithSos(service string, sosServerURL string) error {
+	m := UnRegisterReqJson{service}
+	return makeRequestToServer("POST", sosServerURL+"/unregister", m)
+}
+
+func makeRequestToServer(reqType, url string, reqJson interface{}) error {
+	b, err := json.Marshal(reqJson)
+	if err != nil {
+		return err
+	}
+
+	req, err := http.NewRequest(reqType, url, bytes.NewBuffer(b))
+	if err != nil {
+		return err
+	}
+
+	res, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return err
+	}
+
+	if res.StatusCode < 200 || res.StatusCode >= 300 {
+		decoder := json.NewDecoder(res.Body)
+		defer res.Body.Close()
+		var retMsg struct{ Error string }
+		if err := decoder.Decode(&retMsg); err != nil {
+			return err
+		}
+		if retMsg.Error != "" {
+			return fmt.Errorf(retMsg.Error)
+		}
+	}
+
+	return nil
+}
+
+// StartServiceServer establishes a listener on a random port, registers all neccesary patterns
+// to the router passed in, registers the service with SoS, and starts serving
+// the service on the random port selected before. If any of the above step fails, this function
+// will fail. If the function call needs to know what port that the service is listenning on,
+// they can passed in a uint pointer to the portNumReq to get the port. This function wraps around
```

Comment:
why not just have multiple return values?

Reply:
Because this function starts the server listening on the port, it will never return until there is an error or a signal. That means the port of the function caller would never be set. I can break this function down to two pieces, get the Port number and starts the server.

Reply:
So you'd call this function, and... when is the port filled in? This is fundamentally racy. It's possible you use the port sometime between the call to this function and the *portNumReq = X assignment, and then what?

Reply:
At the moment, the port is only used when the server serves the HTML page, because the JS script needs to know a port to send the request to.
That event only happens after the server is already up and running, which means the port has already been filled in.
But I can see how it can be mis-used in the future, I think I'll go and fix this

Reply:
Fixed